### PR TITLE
Bug 2037274: pkg/*: introduce x509_missing_san_total metric

### PR DESCRIPTION
**Pull request by s-urbaniak**
This introduces a x509_missing_san_total metric for detecting legacy certificates for external identity providers who don't provide a SAN field, see https://github.com/openshift/enhancements/pull/980 for details.

### CODE REVIEW

**Comment by s-urbaniak** [comment]
proof PR seems successful, we have metrics:
image

File: go.mod [hunk 1]
```diff
@@ -10,28 +10,27 @@ require (
...
 	gopkg.in/square/go-jose.v2 v2.2.2
-	k8s.io/api v0.21.0
-	k8s.io/apimachinery v0.21.0
-	k8s.io/apiserver v0.21.0
-	k8s.io/client-go v0.21.0
-	k8s.io/component-base v0.21.0
-	k8s.io/klog/v2 v2.8.0
+	k8s.io/api v0.22.2
+	k8s.io/apimachinery v0.22.2
+	k8s.io/apiserver v0.22.1
```
Comment (stlaz) [inline go.mod:26]: should be 0.22.2

File: pkg/prometheus/metrics.go [hunk 2]
```diff
@@ -15,6 +15,17 @@ const (
 	ErrorResult   = "error"
 )
 
+type Provider string
```
Comment (stlaz) [inline pkg/prometheus/metrics.go:18]: https://github.com/openshift/api/blob/master/config/v1/types_oauth.go#L166

File: pkg/prometheus/metrics.go [hunk 3]
```diff
@@ -77,3 +103,7 @@ func RecordFormPasswordAuth(result string) {
 	authFormCounter.Inc()
 	authFormCounterResult.WithLabelValues(result).Inc()
 }
+
+func X509MissingSANMetric(p Provider) metrics.CounterMetric {
+	return x509MissingSANCounter.WithLabelValues(string(p))
+}
```
Comment (stlaz) [inline pkg/prometheus/metrics.go:109]: inline

File: pkg/prometheus/metrics.go [hunk 4]
```diff
@@ -15,6 +15,17 @@ const (
...
 
+type Provider string
+
+const (
+	ProviderGitHub    Provider = "github"
+	ProviderGitLab             = "gitlab"
+	ProviderGoogle             = "google"
+	ProviderOpenID             = "openid"
+	ProviderBasicAuth          = "basicauth"
+	ProviderKeystone           = "keystone"
```
Comment (stlaz) [inline pkg/prometheus/metrics.go:26]: where's LDAP?

File: go.mod (hunk 1 above)
Comment (s-urbaniak) [reply go.mod:26]: fixed

File: pkg/prometheus/metrics.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:18]: ack, let's use those types 👍

File: pkg/prometheus/metrics.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:109]: ack, let's do, the function is probably not worth it.

File: pkg/prometheus/metrics.go (hunk 4 above)
Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:26]: indeed, where is it ... let me add support for that one

Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:26]: oh i know, we're using ldapclient from library-go and don't have access to the underlying transport.

File: pkg/prometheus/metrics.go [hunk 5]
```diff
@@ -59,11 +80,16 @@ func init() {
...
 	legacyregistry.MustRegister(authBasicCounter)
 	legacyregistry.MustRegister(authBasicCounterResult)
+	legacyregistry.MustRegister(x509MissingSANCounter)
 
 	for _, resultLabel := range []string{SuccessResult, FailResult, ErrorResult} {
 		authBasicCounterResult.WithLabelValues(resultLabel)
 		authFormCounterResult.WithLabelValues(resultLabel)
 	}
+
+	for _, provider := range []Provider{ProviderGitHub, ProviderGitLab, ProviderGoogle, ProviderOpenID, ProviderKeystone, ProviderBasicAuth} {
```
Comment (ibihim) [inline pkg/prometheus/metrics.go:90]: mini-nit:
- `[...]Provider{ProviderGitHub, ProviderGitLab, ProviderGoogle, ProviderOpenID, ProviderKeystone, ProviderBasicAuth}`

**Review by ibihim — APPROVED** [review]
/lgtm

File: pkg/oauthserver/auth.go [hunk 6]
```diff
@@ -563,7 +566,7 @@ func (c *OAuthServerConfig) getPasswordAuthenticator(identityProvider osinv1.Ide
 		if len(connectionInfo.URL) == 0 {
 			return nil, fmt.Errorf("URL is required for BasicAuthPasswordIdentityProvider")
 		}
-		transport, err := transportFor(connectionInfo.CA, connectionInfo.CertInfo.CertFile, connectionInfo.CertInfo.KeyFile)
+		transport, err := transportFor(connectionInfo.CA, connectionInfo.CertInfo.CertFile, connectionInfo.CertInfo.KeyFile, metrics.ProviderBasicAuth)
```
Comment (ibihim) [inline pkg/oauthserver/auth.go:569]: nit: pretty long line.

File: pkg/oauthserver/auth.go [hunk 7]
```diff
@@ -574,7 +577,7 @@ func (c *OAuthServerConfig) getPasswordAuthenticator(identityProvider osinv1.Ide
 		if len(connectionInfo.URL) == 0 {
 			return nil, fmt.Errorf("URL is required for KeystonePasswordIdentityProvider")
 		}
-		transport, err := transportFor(connectionInfo.CA, connectionInfo.CertInfo.CertFile, connectionInfo.CertInfo.KeyFile)
+		transport, err := transportFor(connectionInfo.CA, connectionInfo.CertInfo.CertFile, connectionInfo.CertInfo.KeyFile, metrics.ProviderKeystone)
```
Comment (ibihim) [inline pkg/oauthserver/auth.go:580]: nit: pretty long line

File: pkg/prometheus/metrics.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:90]: i like it, let's make it an array.

File: pkg/prometheus/metrics.go (hunk 4 above)
Comment (s-urbaniak) [reply pkg/prometheus/metrics.go]: Implementation: https://github.com/openshift/oauth-server/blob/2d5618f7908be0006a239905c60af5f009cc9492/pkg/oauthserver/auth.go#L539-L554

Necessary library-go change: https://github.com/openshift/library-go/pull/1278

File: pkg/oauthserver/auth.go (hunk 6 above)
Comment (s-urbaniak) [reply pkg/oauthserver/auth.go]: let's leave this for easier diffing, this code is only present in 4.9.

File: pkg/oauthserver/auth.go (hunk 7 above)
Comment (s-urbaniak) [reply pkg/oauthserver/auth.go]: let's leave this for easier diffing, this code is only present in 4.9.

File: pkg/oauthserver/auth.go [hunk 8]
```diff
@@ -26,6 +27,7 @@ import (
 	"k8s.io/client-go/util/cert"
 	"k8s.io/klog/v2"
 
+	configv1 "github.com/openshift/api/config/v1"
```
Comment (ibihim) [inline pkg/oauthserver/auth.go:30]: pretty cool

File: pkg/prometheus/metrics.go [hunk 9]
```diff
@@ -59,11 +72,24 @@ func init() {
...
 	legacyregistry.MustRegister(authBasicCounter)
 	legacyregistry.MustRegister(authBasicCounterResult)
+	legacyregistry.MustRegister(X509MissingSANCounter)
 
 	for _, resultLabel := range []string{SuccessResult, FailResult, ErrorResult} {
 		authBasicCounterResult.WithLabelValues(resultLabel)
 		authFormCounterResult.WithLabelValues(resultLabel)
 	}
+
+	for _, provider := range [...]configv1.IdentityProviderType{
```
Comment (ibihim) [inline pkg/prometheus/metrics.go:82]: I really don't want to open pandora's box... but that list would fit nicely into the api? :laughing:

Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:82]: i am fine tackling this is in an initiative tracking >=4.10

**Comment by s-urbaniak** [comment]
1.  If the Bugzilla associated with the PR has the "FastFix" keyword, the subjective assessment on the issue has already been done and a customer is impacted. These PRs should be prioritized for merge.
    - [ ] verified
    - [ ] does not apply
2. The bug has significant impact either through severity, reduction in supportability, or number of users affected.
    - [ ] verified
    - [x] does not apply
3. For branches that are in the Maintenance lifecycle phase:
    - [x] The bug is a critical fix, no reasonable workaround exists, and a recommendation for upgrade has been ruled out, or 
    - [ ] The bug is a security related bug
    - [ ] Branch not in maintenance mode yet (current release + previous release for 90 days after current GA; everything older is in maintenance)
4. The severity field of the bug must be set to accurately reflect criticality.
    - [x] verified
5. The PR was created with the cherry-pick bot OR the PR’s description is well formed with user-focused release notes that state the bug number, impact, cause, and resolution. Where appropriate, it should also contain information about how a user can identify whether a particular cluster is affected.
    - [x] verified (here, concretely, we don't want this change to be present in master)

File: pkg/prometheus/metrics.go [hunk 10]
```diff
@@ -15,6 +17,8 @@ const (
 	ErrorResult   = "error"
 )
 
+type Provider string
```
Comment (slaskawi) [inline pkg/prometheus/metrics.go:20]: Do we need this type here? I think you replaced this and this is just a leftover.

Comment (s-urbaniak) [reply pkg/prometheus/metrics.go:20]: indeed, removing

**Comment by xingxingxia** [comment]
Sorry for late sparing time checking on the 4.10 ON_QA bug 2031839 - Starting from Go 1.17 invalid certificates will render a cluster dysfunctional, labeling now:
/label cherry-pick-approved

**Pull request merged**